File: 프로그래머스/3/43163. 단어 변환/단어 변환.py
from collections import deque


def solution(begin, target, words):
    visited = [0]*len(words)
    q = deque()
    q.append((begin,0))
    
    while q:
        nowstr,num = q.popleft()

        if nowstr==target:
            return num
        
        for i in range(len(words)):
            if visited[i]!=0 or words[i]==begin:
                continue
            cnt = 0
            for r, s in zip(words[i], nowstr):
                if r==s:
                    cnt+=1
            if cnt==len(nowstr)-1:
                visited[i]=1
                q.append((words[i],num+1))
    return 0

File: 프로그래머스/3/43163. 단어 변환/단어 변환.cpp
#include <string>
#include <vector>
#include <queue>
#include <map>

using namespace std;

int solution(string begin, string target, vector<string> words) {
    int answer = 0;
    
    
    int visited[50]={0,};
    queue<pair<string, int>> q;
    q.push({begin,0});

    while (q.size()){
        string nowStr = q.front().first;
        int cnt = q.front().second;
        q.pop();

        if (nowStr == target){
            return cnt;
        }

        for (int i=0;i<words.size();i++){
            int diffCnt=0;
            for (int j=0;j<words[i].length();j++){
                if (nowStr[j]!=words[i][j]){
                    diffCnt++;
                }
            }
            if (diffCnt==1 && visited[i]==0){
                visited[i]=1;
                q.push({words[i],cnt+1});
            }
        }

    }
    
    
    return 0;
}